modules/update-restricted-network-cli.adoc

```diff
@@ -0,0 +1,168 @@
+// Module included in the following assemblies:
+//
+// * updating/updating-restricted-network.adoc
+
+[id="update-restricted-network-cli_{context}"]
+= Update the restricted network cluster
+
+After you update the contents of the mirror registry, you can update your
+restricted network cluster.
+
+.Prerequisites
+
+* You have access to the mirror registry that you used to store the images that
+you used to install {product-title}.
+* You have the `ImageContentSourcePolicy` information for the content that you
+mirrored.
+* The the OpenShift Command-line Interface (CLI), commonly known as `oc`, is
+installed.
+* The `jq` package is installed.
+
+.Procedure
+
+Complete the following steps on the bastion host:
+
+. View any existing `ImageContentSourcePolicy` resources for your cluster:
++
+----
+$ oc get ImageContentSourcePolicy -o yaml
+----
+
+. If the previous command returned an `ImageContentSourcePolicy` resource,
+remove it:
++
+----
+$ oc delete ImageContentSourcePolicy <name> <1>
+----
+<1> For `<name>`, specify the name of the existing `ImageContentSourcePolicy` resource.
+
+. Create an `ImageContentSourcePolicy` resource for your cluster:
++
+----
+$ cat test.yaml
+apiVersion: operator.openshift.io/v1alpha1 <1>
+kind: ImageContentSourcePolicy
+metadata:
+  name: example
+spec:
+  repositoryDigestMirrors:
+  - mirrors:
+    - <local_registry_host_name>:<local_registry_host_port>/<repository_name>/release
+    source: quay.io/openshift-release-dev/ocp-release
+  - mirrors:
+    - <local_registry_host_name>:<local_registry_host_port>/<repository_name>/release
+    source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
+----
+<1> Specify the entire specification that was showin in the output of the
+`oc adm mirror release` command that mirrored the images for this release.
+
+. Create the `ImageContentSourcePolicy` resource that you defined:
++
+----
+$ oc create -f test.yaml
+----
+
+////
+. Review the `ImageContentSourcePolicy` resource for your cluster:
++
+----
+$ oc edit ImageContentSourcePolicy
+
+apiVersion: operator.openshift.io/v1alpha1
+kind: ImageContentSourcePolicy
+metadata:
+  name: example
+spec:
+  repositoryDigestMirrors: <1>
+  - mirrors:
+    - <local_registry_host_name>:<local_registry_host_port>/<repository_name>/release
+    source: quay.io/openshift-release-dev/ocp-release
+  - mirrors:
+    - <local_registry_host_name>:<local_registry_host_port>/<repository_name>/release
+    source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
+----
+<1> The `repositoryDigestMirrors` section must contain the `mirrors` data that
+is in the output of the `oc adm mirror release` command from when you mirrored
+the images for this release.
+
+. If the `repositoryDigestMirrors` section of the `ImageContentSourcePolicy`
+does not contain the `mirrors` for this release, add them to that section.
+////
+. Start the cluster update:
```

Comment:
If new ImageContentSourcePolicy is created, we need do some more check before start the cluster update. 
The new ImageContentSourcePolicy settings will be deployed to each node and shortly start using the mirrored repository for requests to the source repository.

To check that the mirrored configuration worked, go to one of your nodes. For example:
`
$ oc get node

NAME                         STATUS ROLES AGE VERSION

ip-10-0-137-44.ec2.internal  Ready worker 7m v1.14.6+90fadebfa

ip-10-0-138-148.ec2.internal Ready          master 11m v1.14.6+90fadebfa

ip-10-0-139-122.ec2.internal Ready          master 11m v1.14.6+90fadebfa

ip-10-0-147-35.ec2.internal  Ready,SchedulingDisabled       worker 7m v1.14.6+90fadebfa

ip-10-0-153-12.ec2.internal  Ready worker 7m v1.14.6+90fadebfa

ip-10-0-154-10.ec2.internal  Ready master 11m v1.14.6+90fadebfa

`

You can see that scheduling on each worker node is disabled as the change is being applied.

If no ImageContentSourcePolicy resource is created, we could skip this check.

Reply:
> If new ImageContentSourcePolicy is created, we need do some more check before start the cluster update...

Ugh.  What if you miss a SchedulingDisabled?  This really needs to be represented in `status` on some object so we can use a level-check instead of SchedulingDisabled edge checks.

Reply:
Hmm, I am not sure what unexpected behaviour will happen when upgrade is started but new mirror registry configuration is not laid down onto node yet. Maybe upgrade will reconcile until new mirror registry configuration is totally laid down. I think waiting for new ImageContentSourcePolicy setting onto nodes before upgrade would be safer.

Reply:
I agree that upgrading after mirror propagation is safer.  I just wish we had something better than hoping to catch all the SchedulingDisabled as they flit past.

Reply:
okay. If use other way instead of SchedulingDisabled checks, I would suggest to use `oc describe node | grep machineconfig` to ensure each node is in 'Done' state, that means new ImageContentSourcePolicy mirror setting is applied to node via MCO, WDYT?

Reply:
> ...to ensure each node is in 'Done' state...

But presumably it was in Done before you added the new ImageContentSourcePolicy.  I don't know what sort of delay there is before you start seeing node-side effects.  And there's [no `status` on ImageContentSourcePolicy][1]...  I'll see if I can get something more clear, but checking for `Done` status on the nodes after a brief wait seems like the best we have at the moment.

[1]: https://github.com/openshift/api/blob/9525304a0adb725ab4a4a54539a1a6bf6cc343d3/operator/v1alpha1/types_image_content_source_policy.go#L9-L20